### ISSUE TITLE
invalid value for BigDecimal(): "-e1535637460290" (ArgumentError)

### ISSUE THREAD

**Issue by dentarg** [ohler55/oj#503]
Repro:

```
$ ruby -v
ruby 2.4.4p296 (2018-03-28 revision 63013) [x86_64-darwin16]

$ gem install oj
Fetching: oj-3.6.10.gem (100%)
Building native extensions. This could take a while...
Successfully installed oj-3.6.10
Parsing documentation for oj-3.6.10
Installing ri documentation for oj-3.6.10
Done installing documentation for oj after 2 seconds
1 gem installed

$ ruby -r oj -e 'Oj.load("{\"number\":-e1535637460290}")'
-e:1:in `BigDecimal': invalid value for BigDecimal(): "-e1535637460290" (ArgumentError)
	from -e:1:in `load'
	from -e:1:in `<main>'
```

For reference, this is not a problem using Ruby 2.3:

```
$ ruby -v
ruby 2.3.7p456 (2018-03-28 revision 63024) [x86_64-darwin16]

$ gem install oj
Fetching: oj-3.6.10.gem (100%)
Building native extensions. This could take a while...
Successfully installed oj-3.6.10
Parsing documentation for oj-3.6.10
Installing ri documentation for oj-3.6.10
Done installing documentation for oj after 3 seconds
1 gem installed

$ ruby -r oj -e 'Oj.load("{\"number\":-e1535637460290}")'

$ echo $?
0
```

I think the problem is related to https://github.com/ruby/bigdecimal/issues/47, https://github.com/kubo/ruby-oci8/pull/144

```
$ chruby-exec 2.3 -- ruby -r bigdecimal -e 'puts BigDecimal("-e1535637460290").inspect'
#<BigDecimal:7f9090a31330,'-0.0',9(9)>

$ chruby-exec 2.4 -- ruby -r bigdecimal -e 'puts BigDecimal("-e1535637460290").inspect'
-e:1:in `BigDecimal': invalid value for BigDecimal(): "-e1535637460290" (ArgumentError)
	from -e:1:in `<main>'
```

**Comment by ohler55** [comment]
I'm not sure I understand why this is a problem. `-e123` is not a valid scientific notation so why would you expect it to not generate an error. If Ruby 2.3 accepted that format it probably should not have and with 2.4 that oversight has been corrected. In any case, a tightening up of Ruby in regard to parsing decimals is not an issue with Oj but rather a reflection of the Ruby version being used.

**Comment by dentarg** [comment]
Sorry, I should have been more clear in the original report. I'm reporting the issues because I expect `Oj::ParseError` to be raised instead of `ArgumentError`.

`ruby -r json -e 'JSON.parse("{\"number\":-e1535637460290}")'` raises `JSON::ParserError` (on both 2.3 and 2.4).

**Comment by ohler55** [comment]
I'll take a look but I suspect that doing the setup to catch and re-raise a new exception is going to incur a performance penalty. I'd rather have the ruby error raised if that is the case. Let me check though.

**Comment by dentarg** [comment]
Thanks @ohler55. Seems to work fine for `Oj.load`, I can see it raising `Oj::ParseError` now. 

That doesn't seem to be the case when `Oj::ScHandler` is used. See https://github.com/dentarg/gists/tree/master/gists/oj

**Comment by ohler55** [comment]
oh yeah, the other parsers too. On it.

**Comment by ohler55** [comment]
Ok, pushed a new version.

**Comment by ohler55** [comment]
Released.

**Comment by dentarg** [comment]
Thanks! I think the issue can be closed now.

**Comment by dentarg** [comment]
@ohler55 seems like there's been an regression here

```ruby
$ irb -roj
irb(main):001:0> RUBY_VERSION
=> "2.6.1"
irb(main):002:0> Oj::VERSION
=> "3.7.8"
irb(main):003:0> json = "{\"number\":-e1550121980306}"
=> "{\"number\":-e1550121980306}"
irb(main):004:0> Oj.load(json)
Traceback (most recent call last):
        6: from /Users/dentarg/.rubies/ruby-2.6.1/bin/irb:23:in `<main>'
        5: from /Users/dentarg/.rubies/ruby-2.6.1/bin/irb:23:in `load'
        4: from /Users/dentarg/.rubies/ruby-2.6.1/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        3: from (irb):4
        2: from (irb):4:in `load'
        1: from (irb):4:in `BigDecimal'
ArgumentError (invalid value for BigDecimal(): "-e1550121980306")
```

It did work in an earlier version:

```ruby
$ irb -roj
irb(main):001:0> RUBY_VERSION
=> "2.6.1"
irb(main):002:0> Oj::VERSION
=> "3.6.11"
irb(main):003:0> json = "{\"number\":-e1550121980306}"
=> "{\"number\":-e1550121980306}"
irb(main):004:0> Oj.load(json)
Traceback (most recent call last):
        5: from /Users/dentarg/.rubies/ruby-2.6.1/bin/irb:23:in `<main>'
        4: from /Users/dentarg/.rubies/ruby-2.6.1/bin/irb:23:in `load'
        3: from /Users/dentarg/.rubies/ruby-2.6.1/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):4
        1: from (irb):4:in `load'
Oj::ParseError (Hash/Object not terminated (number) at line 1, column 25 [parse.c:990])
```

**Comment by ohler55** [comment]
I'll look into it. I thought I added a unit test for that. Must not have though.

**Comment by ohler55** [comment]
Fixed but not yet released.

**Comment by ohler55** [comment]
Released as Oj 3.7.9.

**Comment by dentarg** [comment]
Thank you @ohler55, it works as expected
